sdk/Lusid.Sdk.Tests/Utilities/ApiRetryHandlerTests.cs

```diff
@@ -0,0 +1,107 @@
+            testApiConfig.ApiUrl = ListenerUriPrefix;
+            
+            _apiFactory = new LusidApiFactory(testApiConfig);
+        }
+
+        [Test]
+        [Explicit("This passes locally, but not in the pipeline, and no idea why. Setting to explicit until resolved.")]
+        public void CallGetPortfoliosApi_WhenPollyRetryPolicyConfigured_PollyIsTriggered()
+        {
+            const int expectedNumberOfRetries = ApiRetryHandler.MaxRetryAttempts;
+            _httpListener.Start();
+            for (var i = 0; i < expectedNumberOfRetries + 1; i++)
+            {
+                 _httpListener.BeginGetContext(result =>
+                {
+                    var listener = (HttpListener) result.AsyncState;
+                    // Call EndGetContext to complete the asynchronous operation.
+                    var context = listener.EndGetContext(result);
+
+                    // Obtain a response object.
+                    var response = context.Response;
+
+                    // Abort the response. This returns 0 status code.
+                    response.Abort();
+                }, _httpListener);
+            }
+            
+            var currentRetry = 0;
+            var testRetryPolicy = Policy
+                .HandleResult<IRestResponse>(restResponse => restResponse.StatusCode == 0)
+                .Retry(
+                    retryCount: expectedNumberOfRetries,
+                    (result, i) =>
+                    {
+                        Console.WriteLine($"If you see this, then polly retry works. Retry number: {i}");
+                        currentRetry = i;
+                    });
+
+            RetryConfiguration.RetryPolicy = testRetryPolicy;
+
+            // Calling GetPortfolio or any other API triggers the flow that triggers polly
+            var sdkResponse = _apiFactory.Api<IPortfoliosApi>().GetPortfolio("any", "any");
+            
+            Assert.That(currentRetry, Is.EqualTo(expectedNumberOfRetries));
+            // In the future 0 error codes with throw an error after retries exceeded
+            Assert.That(sdkResponse, Is.Null);
+        }
+
+        [Test]
+        public void CreateLusidFactory_WhenRetryPolicyIsNull_AssignsDefaultRetryPolicy()
+        {
+
+            RetryConfiguration.RetryPolicy = null;
+            
+            var newFactory = new LusidApiFactory(new Configuration());
+
+            Assert.That(RetryConfiguration.RetryPolicy, Is.Not.Null);
+        }
+        
+        [Test]
+        public void CreateLusidFactory_WhenRetryPolicyIsAlreadyAssigned_ExistingRetryPolicyIsUsed()
+        {
+            var testPolicy = Policy.HandleResult<IRestResponse>(response => true).Retry();
+            
+            RetryConfiguration.RetryPolicy = testPolicy;
+            var newFactory = new LusidApiFactory(new Configuration());
+
+            Assert.That(RetryConfiguration.RetryPolicy, Is.EqualTo(testPolicy));
+        }
+
```

Comment:
There isn't a single test here that uses the `ApiRetryHandler`. How do you know that it works?

Reply:
We needs tests which show what happens when:

- There is a 2xx response (it's not clear what happens here currently, prove that we don't retry in this situation)
- There is a non-2xx response (it's not clear what happens here currently, what is your intention with this change?)
- We reach the maximum number of retries (it's not clear what happens here currently, what is your intention with this change?)
- If we can't deserialsie the response  (it's not clear what happens here currently, what is your intention with this change?)

We also need a clear understanding of:

- Whether instantiating an ApiClientFactory affects the global RetryConfiguration and whether or not this is the desired behaviour
- The wrap behaviour that you mentioned and how a client would implement that

Reply:
ApiRetryHandler itself isn't really being tested. I'm also not sure how we could or should test it. If we were testing it, then we'd only be testing the Polly library functionality, but open to suggestions here. What we can test though, is to show that when we assign a Polly retry policy, then it does actually work on our selected errors thrown by the REST client

Reply:
I agree on the test cases you suggested, I'll look into adding those.

I can also add an example how Polly wrap can be used to combine policies I guess.

> Whether instantiating an ApiClientFactory affects the global RetryConfiguration and whether or not this is the desired behaviour
Do you mean LusidApiFactory. Is ApiClientFactory separate? Also the desired behaviour is probably up for discussion

Reply:
> If we can't deserialsie the response (it's not clear what happens here currently, what is your intention with this change?)

Open API doesn't currently support retries on deserialization errors by the looks of it. On those, we can implement ExceptionFactoryHandler and throw if this is the case. 

>

Reply:
Yes I meant the `LusidApiFactory`. 

Sounds good regarding the ExceptionFactoryHandler.

I think it is important to test that the `PollyApiRetryHandler` works as expected by actually using it and then asserting the results are as they should be. I can see we have some in-line Polly policies that is testing the Polly library more than it would be if we were to test our `PollyApiRetryHandler` class. 

It makes sense to have some examples on how a client can implement their own Polly handler, however if we are looking at this from the perspective of test coverage what we want to do is ensure that our `PollyApiRetryHandler` does what we expect. It isn't much value to set up an in-line Polly policy that retries on `502` as what is that showing? We want to use our retry handler and demonstrate how it handles a `502` error as every user of the SDK is going to be using the `PollyApiRetryHandler` by default going forward if they use the `LusidApiFactory`.

Reply:
While it looks like it is testing Polly functionality, it's also testing that the behaviour remains as expected whenever we make any changes to default templates or upgrade the SDK.

Also because the way Polly works relies on OpenAPI generated code, it might not behave as someone would expect. One example is how lack of fallback policy gives you a null result rather than to rethrow the error it's retrying on. So these tests give all the necessary examples showing how Polly should properly be implemented

Reply:
Okay I understand, that's great. We should make sure that it is not at the expense of testing the `PollyApiRetryHandler` in its entirety. If you want to count invocations create a variable for counting retries and increment it in `GetHttpResponseHandler` by removing `static` so it can access the variable.

Reply:
genius

Reply:
Done

Reply:
Awesome, nice work!